### DISCUSSION TITLE
Neuralnet on IOS/Android

### DISCUSSION THREAD

**Discussion by Myles-Kel** [opentrack/opentrack#1664] [Ideas]
Instead of running Neuralnet on the PC with it's webcam, maybe we can run it on the phone (which has a superior camera in both resolution and framerate) and send the tracking data to opentrack via WiFi or USB?
P.S. sorry if I'm using the worng terms but I'm not very tech savvy.

**Comment by sthalik** [comment]
What's the CPU of your phone?

**Reply by Myles-Kel** [reply]
Snapdragon 835. Can it not handle Neuralnet? or is it something else?

**Reply by sthalik** [reply]
You'd need someone who:

- knows how to port and cross-compile opentrack for Android
- is familiar with Android APIs such as the foreign function interface, permissions, background services
- wants to learn opentrack internals

There are going to be unforeseen issues, and it's going to take months. How many months of doing it in your spare time will it take? I don't know. Two is the rather optimistic variant.

I don't think it's going to make the phone overheat. It's going to take up to a half of the cores and even with SIMD it won't damage or crash the phone.

Conclusion: get a cheap used laptop instead, like a Sandy Bridge i3/i5 or something.  Then run it using all of that laptop's CPU cores. Just use a wired connection for network access between those 2 computers.

**Reply by Myles-Kel** [reply]
The inspiration behind my idea was the lack of an app that turns my Android into a webcam that has **high fps** and **low latency**. Then I thought to myself: instead of streaming the huge cam footage, why not process the feed right on the phone and then stream the final head pose to the pc (well now I know why thanks to you).

Funny enough, I have an old 2nd gen i5 laptop laying around. Time for it to come out of retirement :)

**Reply by sthalik** [reply]
You can use USB tethering to reduce latency. It's not too bad then. I used the [SoundWire](https://play.google.com/store/apps/details?id=com.georgie.SoundWireFree&hl=en_US&gl=US&pli=1) software and it was pretty usable that way.

You only want to evaluate the model, not evne compute the pose itself since all that takes less than a millisecond -- opentrack runs at 250 Hz for all trackers.

**Comment by DaWelter** [comment]
At least one solution exists already which fits your use case. It's an app called Smoothtrack. Another one is FacePoseApp. The website says it's from the creator of FTnoIR :-O

I guess they are pretty good too. You'll certainly save the CPU load for running the AI.

The idea behind the NN-Tracker was to have that built into OpenTrack for highest convenience. Considering all that, from my point of view it precludes creation of another phone based tracker.

**Reply by Myles-Kel** [reply]
I know of these apps, but none of them so far matches the tracking quality of Neuralnet. Taht's why I was curious to see if porting it to android is possible.

**Reply by DaWelter** [reply]
Thanks for the positive feedback. Appreciated.

If someone wants to use the model, no problem. @sthalik , wasn't quite right. You'd need someone to write the equivalent of the nn-tracker plugin. So, loading the model, grabbing the video feed, performing the inference and sending the result over the network. It might actually not be that hard for an experienced mobile developer.
